### PR TITLE
Unfreeze layers when recovering with slanted triangular learning rate

### PR DESCRIPTION
**Pull request by Hyperparticle**
`is_first_epoch` will always be true for one epoch regardless of the current epoch number. Instead, we should base the unfreezing on the epoch number to unfreeze the correct number of layers when recovering a training run.

Alternatively, could simplify the condition to

```python
num_layers_to_unfreeze = epoch + 2 if epoch > -1 else 1
```

### CODE REVIEW

**Comment by Hyperparticle** [comment]
This may not change what I hoped. I think I had some confusion on the order of operations. My intention was to be able to change the existing `config.json` and have it override the saved training state. Looks like it's currently the other way around, i.e., the checkpointer overrides the config specification (for slanted triangular learning rate, anyways). Is there a way to override the training state when recovering?

**Comment by Hyperparticle** [comment]
I did some more digging. Indeed, the checkpointer overrides any state in the learning rate scheduler, which will cause `slanted_triangular` to always freeze all layers but the last for the first epoch (when `gradual_unfreezing` is enabled), even when recovering. Any easy way to fix this?

**Comment by tingkai-zhang** [comment]
Hi, thanks for finding the problem and gave the fix! Did you check this issue? #2388 
I thought the problem stated in the comment is similar to what is described here, isn't it?

**Comment by Hyperparticle** [comment]
@tingkai-zhang It's similar, but I'm more concerned about changing the configuration of LR schedulers when recovering from a previous run. The strange behavior you see with slanted triangular is intentional, but I'm not sure if it really helps.

**Comment by matt-gardner** [comment]
@matt-peters, can you take a look at this and decide what the best course of action is here?

**Comment by matt-peters** [comment]
@Hyperparticle Will adding `state_dict` and `load_state_dict` methods to `slanted_triangular` to persist the state at in the checkpoint accomplish what you want?

**Comment by Hyperparticle** [comment]
@matt-peters It appears the state is already persisted across checkpoints, just incorrectly. The other problem is there is currently no easy way (I think) to override that state by changing the config file when resuming training.

**Comment by matt-peters** [comment]
I'd suggest fixing both problems by: (a) adding the correct state to the checkpoint so it's available if the user doesn't want to override (b) adding a parameter to the constructor that allows the user to override the checkpoint state if desired.

**Comment by Hyperparticle** [comment]
Sounds reasonable. I'll try that out.

**Comment by Hyperparticle** [comment]
I've gone with a custom solution, so I'm closing this.
